service/storage_service.cc

```diff
@@ -814,6 +814,26 @@ void storage_service::join_token_ring(int delay) {
         // If we're replacing, _cdc_streams_ts should have been set in prepare_to_join.
         assert(!db().local().is_replacing());
 
+        if (!db::system_keyspace::bootstrap_complete() /* second case */) {
+            // Check if we managed to learn about other nodes' tokens (if there are other nodes).
+            // We might have not due to misconfiguration, e.g. skipping the "wait for gossip to settle" phase,
+            // or due to a network partition which didn't allow us to contact other seeds.
+            // But CDC requires that we know other nodes' tokens - they are needed to make a new generation of streams.
+            const auto& seeds = _gossiper.get_seeds();
+            if (!seeds.count(get_broadcast_address()) || seeds.size() > 1) {
+                auto& tok_to_ep = _token_metadata.get_token_to_endpoint();
+                if (std::none_of(tok_to_ep.begin(), tok_to_ep.end(),
+                        [me = get_broadcast_address()] (const std::pair<token, inet_address> p) { return p.second != me; })) {
+                    throw std::runtime_error(format(
+                        "Can't boot this node because we weren't able to retrieve other nodes' tokens,"
+                        " which is required for CDC to work. Make sure that:\n"
```

Comment:
If you claim that this is needed just for CDC to work, then it shouldn't be an error if CDC is disabled.
Ok, maybe it is always a problem and not just for CDC? How can normal Scylla (without CDC) function without knowing any other nodes' tokens, and yet this node is not a seed?

Reply:
> If you claim that this is needed just for CDC to work, then it shouldn't be an error if CDC is disabled.

And it won't: it's protected by `db().local().get_config().check_experimental(db::experimental_features_t::CDC)`.

>  Ok, maybe it is always a problem and not just for CDC? How can normal Scylla (without CDC) function without knowing any other nodes' tokens, and yet this node is not a seed?

This code path is entered in two cases only:
1. seed
2. not seed, but auto_bootstrap=off
if the user uses auto_bootstrap=off, they should know what they are doing.

A Scylla node can function without knowing others' tokens, it will just send writes to the wrong replicas (to itself). The user is able to shoot themselves in the foot if they really want (using options like auto_bootstrap=off or skip-wait-for-gossip-to-settle).

I would like to give the users less power so developers have more power (and when it's better for the users, even if they don't know it), but the options are already there, inherited from Cassandra, and I can't do anything about it. But I can disallow using these options if they want to use the new shiny feature that I'm working on :>